File: app/test/transfer_crypto_test.dart
import 'dart:math';

import 'package:flutter_test/flutter_test.dart';
import 'package:vidyut/src/shared/transfer_crypto.dart';

void main() {
  const metadata = TransferChunkMetadata(
    transferId: 'transfer_1234567890',
    fileId: 'file_1234567890123',
    offset: 4096,
    plaintextBytes: 5,
  );

  group('transfer chunk crypto', () {
    test('round-trips an independently authenticated chunk', () async {
      final crypto = TransferCrypto(random: Random(7));
      final chunk = await crypto.encrypt(
        metadata: metadata,
        plaintext: [1, 2, 3, 4, 5],
        pairingSecret: 'pairing-secret',
      );

      expect(
        await crypto.decrypt(chunk: chunk, pairingSecret: 'pairing-secret'),
        [1, 2, 3, 4, 5],
      );
    });

    test('binds ciphertext to its offset', () async {
      final crypto = TransferCrypto(random: Random(7));
      final chunk = await crypto.encrypt(
        metadata: metadata,
        plaintext: [1, 2, 3, 4, 5],
        pairingSecret: 'pairing-secret',
      );
      final tampered = EncryptedTransferChunk(
        transferId: chunk.transferId,
        fileId: chunk.fileId,
        offset: 0,
        plaintextBytes: chunk.plaintextBytes,
        nonce: chunk.nonce,
        ciphertext: chunk.ciphertext,
      );

      expect(
        () => crypto.decrypt(chunk: tampered, pairingSecret: 'pairing-secret'),
        throwsA(isA<Object>()),
      );
    });

    test('rejects mismatched plaintext size before encryption', () {
      final crypto = TransferCrypto();

      expect(
        () => crypto.encrypt(
          metadata: metadata,
          plaintext: [1, 2],
          pairingSecret: 'pairing-secret',
        ),
        throwsFormatException,
      );
    });

    test('produces a lowercase whole-file SHA-256', () async {
      final hash = await TransferCrypto().sha256Hex([97, 98, 99]);

      expect(
        hash,
        'ba7816bf8f01cfea414140de5dae2223'
        'b00361a396177a9cb410ff61f20015ad',
      );
    });
  });
}

File: tests/transfer-crypto.test.ts
import { describe, expect, test } from "bun:test";
import {
  decryptTransferChunk,
  encryptTransferChunk,
  sha256Hex,
} from "../src/shared/transfer-crypto";

const metadata = {
  transferId: "transfer_1234567890",
  fileId: "file_1234567890123",
  offset: 4096,
  plaintextBytes: 5,
};

describe("transfer chunk crypto", () => {
  test("round-trips an independently authenticated chunk", async () => {
    const plaintext = new Uint8Array([1, 2, 3, 4, 5]);
    const chunk = await encryptTransferChunk(
      metadata,
      plaintext,
      "pairing-secret",
      new Uint8Array(12).fill(7),
    );

    await expect(
      decryptTransferChunk(chunk, "pairing-secret"),
    ).resolves.toEqual(plaintext);
  });

  test("binds ciphertext to transfer, file, offset and byte count", async () => {
    const chunk = await encryptTransferChunk(
      metadata,
      new Uint8Array([1, 2, 3, 4, 5]),
      "pairing-secret",
    );

    await expect(
      decryptTransferChunk({ ...chunk, offset: 0 }, "pairing-secret"),
    ).rejects.toThrow();
    await expect(
      decryptTransferChunk(
        { ...chunk, fileId: "file_DIFFERENT_1234" },
        "pairing-secret",
      ),
    ).rejects.toThrow();
  });

  test("rejects mismatched plaintext size before encryption", async () => {
    await expect(
      encryptTransferChunk(
        metadata,
        new Uint8Array([1, 2]),
        "pairing-secret",
      ),
    ).rejects.toThrow(/byte count/);
  });

  test("produces a lowercase whole-file SHA-256", async () => {
    await expect(sha256Hex(new TextEncoder().encode("abc"))).resolves.toBe(
      "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad",
    );
  });
});

